python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -0,0 +1,226 @@
+
+    return typer
+
+
+@register_model(GroupType)
+class GroupModel(models.StructModel):
+    def __init__(
+        self, dmm, fe_type
+    ):  # fe_type is fully instantiated group type
+        members = [
+            ("group_data", types.CPointer(fe_type.group_scalar_type)),
+            ("size", types.int64),
+            ("index", types.CPointer(fe_type.index_type)),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+call_cuda_functions: Dict[Any, Any] = {}
+
+
+def _register_cuda_reduction_caller(func, inputty, retty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}", retty(types.CPointer(inputty), types.int64)
+    )
+
+    def caller(data, size):
+        return cuda_func(data, size)
+
+    if call_cuda_functions.get(func.lower()) is None:
+        call_cuda_functions[func.lower()] = {}
+
+    type_key = (retty, inputty)
+    call_cuda_functions[func.lower()][type_key] = caller
+
+
+def _register_cuda_idxreduction_caller(func, inputty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}",
+        types.int64(
+            types.CPointer(inputty), types.CPointer(types.int64), types.int64
+        ),
+    )
+
+    def caller(data, index, size):
+        return cuda_func(data, index, size)
+
+    # only support default index type right now
+    type_key = (index_default_type, inputty)
+    if call_cuda_functions.get(func.lower()) is None:
+        call_cuda_functions[func.lower()] = {}
+    call_cuda_functions[func.lower()][type_key] = caller
+
+
+def _create_reduction_attr(name, retty=None):
+    class Attr(AbstractTemplate):
+        key = name
+
+    if retty:
+
+        def generic(self, args, kws):
+            return nb_signature(retty, recvr=self.this)
+
+    else:
+
+        def generic(self, args, kws):
+            return nb_signature(self.this.group_scalar_type, recvr=self.this)
+
+    Attr.generic = generic
+
+    def _attr(self, mod):
+        return types.BoundFunction(
+            Attr, GroupType(mod.group_scalar_type, mod.index_type)
+        )
+
+    return _attr
+
+
+class GroupIdxMax(AbstractTemplate):
+    key = "GroupType.idxmax"
+
+    def generic(self, args, kws):
+        return nb_signature(self.this.index_type, recvr=self.this)
+
+
+class GroupIdxMin(AbstractTemplate):
+    key = "GroupType.idxmin"
+
+    def generic(self, args, kws):
+        return nb_signature(self.this.index_type, recvr=self.this)
+
+
+@cuda_registry.register_attr
+class GroupAttr(AttributeTemplate):
+    key = GroupType
+
+    def resolve_idxmax(self, mod):
+        return types.BoundFunction(
+            GroupIdxMax, GroupType(mod.group_scalar_type, mod.index_type)
+        )
+
+    def resolve_idxmin(self, mod):
+        return types.BoundFunction(
+            GroupIdxMin, GroupType(mod.group_scalar_type, mod.index_type)
+        )
+
+
+setattr(GroupAttr, "resolve_max", _create_reduction_attr("GroupType.max"))
```

Comment:
Why not just say:

```
    resolve_max = _create_reduction_attr("GroupType.max")
    ...

```

In the definition of `GroupAttr` above?

Reply:
Done! :)